src/zenml/integrations/mlflow/steps/mlflow_deployer.py

```diff
@@ -82,79 +85,91 @@ def mlflow_deployer_step(
-                pipeline_name=step_env.pipeline_name,
-                step_name=step_env.step_name,
-                step_context=context,
-            ),
-        )
-        if last_service and not isinstance(
-            last_service, MLFlowDeploymentService
-        ):
-            raise ValueError(
-                f"Last service deployed by step {step_env.step_name} and "
-                f"pipeline {step_env.pipeline_name} has invalid type. Expected "
-                f"MLFlowDeploymentService, found {type(last_service)}."
+        pipeline_name = step_env.pipeline_name
+        run_id = step_env.pipeline_run_id
+        step_name = step_env.step_name
+
+        # get the MLflow model deployer stack component
+        mlflow_model_deployer = Repository().active_stack.model_deployer
+        if not isinstance(mlflow_model_deployer, MLFlowModelDeployer):
+            raise TypeError(
+                f"Expected ModelDeployer type MLFlowModelDeployer but got "
+                f"{type(mlflow_model_deployer)} instead"
             )
 
-        mlflow_step_env = cast(
-            MLFlowStepEnvironment, Environment()[MLFLOW_STEP_ENVIRONMENT_NAME]
+        # fetch existing services with same pipeline name, step name and model name
+        existing_services = mlflow_model_deployer.find_model_server(
+            pipeline_name=pipeline_name,
+            step_name=step_name,
+            model_name=config.model_name,
         )
-        client = MlflowClient()
-        # fetch the MLflow artifacts logged during the pipeline run
-        model_uri = None
-        mlflow_run = mlflow_step_env.mlflow_run
-        if mlflow_run and client.list_artifacts(
-            mlflow_run.info.run_id, config.model_name
-        ):
-            model_uri = get_artifact_uri(config.model_name)
-
-        if not model_uri:
+
+        # check for conditions to deploy the model
+        if not config.model_uri:
             # an MLflow model was not found in the current run, so we simply reuse
             # the service created during the previous step run
-            if not last_service:
+            if not existing_services:
                 raise RuntimeError(
                     f"An MLflow model with name `{config.model_name}` was not "
                     f"trained in the current pipeline run and no previous "
                     f"service was found."
                 )
-            return last_service
-
-        if not deploy_decision and last_service:
-            return last_service
-
-        # stop the service created during the last step run (will be replaced
-        # by a new one to serve the new model)
-        if last_service:
-            last_service.stop(timeout=10)
+            return existing_services[0]
+
+        if not deploy_decision:
+            if existing_services:
+                return existing_services[0]
+            else:
+                # TODO: investigate what to do in this case. For now, returning 
```

Comment:
this TODO doesn't represent what's happening below. Either way, this is a problem I've been dealing with too: what to return if the deploy_decision is False and there is no previously running service for this pipeline/step/model ? Two solutions I could think of:

1. raise an error, like you did here. But this doesn't really make sense from an end user perspective, right ? For a user running a continuous deployment pipeline, this is a valid use-case.
2. deploy the service anyway, even if the decision is False. We can justify this by saying that we always need a baseline service running.
3. (my personal favorite) just create and return an inactive service:

```
service = MLFlowDeploymentService(config)
return service
```

Reply:
This reminds me that there is also an inference pipeline in our mlflow_deployment example and a post-execution workflow that both make use of the `load_last_service_from_step` utility and should be updated to use the new MLflow model deployer API instead.

Reply:
Thanks for the suggestions. I like the 3rd option best too and will return an inactive service like I was doing before.

Reply:
thanks for pointing out the example. I'll update it